File: module/storage-kv/db_lease.go
package main

import (
	"encoding/binary"

	"github.com/pantopic/ext-mdb/sdk-go"
)

type dbLeaseImpl struct {
	db
}

func (db dbLeaseImpl) init(txn mdb.Txn) {
	db.open(txn)
}

func (db dbLeaseImpl) get(txn mdb.Txn, id uint64) (item lease, err error) {
	k := binary.AppendUvarint(nil, id)
	var v []byte
	v, err = txn.Get(db.i, k, v)
	if mdb.IsNotFound(err) {
		err = nil
		return
	}
	if err != nil {
		return
	}
	return item.FromBytes(k, v)
}

func (db dbLeaseImpl) put(txn mdb.Txn, item lease) error {
	return txn.Put(db.i, binary.AppendUvarint(nil, item.id), item.Bytes(nil), 0)
}

func (db dbLeaseImpl) all(txn mdb.Txn) (items []lease, err error) {
	cur, err := txn.OpenCursor(db.i)
	if err != nil {
		return nil, err
	}
	defer cur.Close()
	var item lease
	var k, v []byte
	k, v, err = cur.Get(k, v, mdb.Next)
	for !mdb.IsNotFound(err) && len(k) > 0 {
		if err != nil {
			return nil, err
		}
		item, err := item.FromBytes(k, v)
		if err != nil {
			return nil, err
		}
		items = append(items, item)
		k, v, err = cur.Get(k[:0], v[:0], mdb.Next)
	}
	if mdb.IsNotFound(err) {
		err = nil
	}
	return
}

func (db dbLeaseImpl) del(txn mdb.Txn, id uint64) error {
	return txn.Del(db.i, binary.AppendUvarint(nil, id), nil)
}

File: module-zig/storage-kv/db_lease.zig
//! Mirrors module/storage-kv/db_lease.go

const std = @import("std");
const mdb = @import("mdb");

const Db = @import("db.zig").Db;
const Lease = @import("lease.zig").Lease;
const util = @import("util.zig");

pub const DbLease = struct {
    db: Db,

    pub fn init(self: DbLease, txn: mdb.Txn) void {
        self.db.open(txn);
    }

    pub fn get(self: DbLease, txn: mdb.Txn, id: u64) !Lease {
        var kbuf: [util.max_varint_len]u8 = undefined;
        const k = kbuf[0..util.putUvarint(&kbuf, id)];
        const v = txn.get(self.db.i, k) catch |err| {
            if (err == mdb.Error.NotFound) {
                return Lease{};
            }
            return err;
        };
        return Lease.fromBytes(k, v);
    }

    pub fn put(self: DbLease, txn: mdb.Txn, item: Lease) !void {
        var kbuf: [util.max_varint_len]u8 = undefined;
        const k = kbuf[0..util.putUvarint(&kbuf, item.id)];
        var vbuf: [2 * util.max_varint_len + 4]u8 = undefined;
        return txn.put(self.db.i, k, item.bytes(&vbuf), 0);
    }

    pub fn all(self: DbLease, txn: mdb.Txn, allocator: std.mem.Allocator) ![]Lease {
        var items = std.ArrayList(Lease).empty;
        const cur = try txn.openCursor(self.db.i);
        defer cur.close();
        while (true) {
            const entry = cur.get("", "", mdb.op_next) catch |err| {
                if (err == mdb.Error.NotFound) break;
                return err;
            };
            if (entry.key.len == 0) break;
            const item = try Lease.fromBytes(entry.key, entry.val);
            try items.append(allocator, item);
        }
        return items.items;
    }

    pub fn del(self: DbLease, txn: mdb.Txn, id: u64) !void {
        var kbuf: [util.max_varint_len]u8 = undefined;
        const k = kbuf[0..util.putUvarint(&kbuf, id)];
        return txn.del(self.db.i, k, "");
    }
};
